flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/adaptive/CreatingExecutionGraphTest.java

```diff
@@ -155,6 +179,60 @@ public void testSuccessfulSlotAssignmentTransitionsToExecuting() throws Exceptio
         }
     }
 
+    @Test
+    public void testOperatorCoordinatorUsesFailureHandlerOfTheCurrentState() throws Exception {
+        try (MockCreatingExecutionGraphContext context = new MockCreatingExecutionGraphContext()) {
+            final CompletableFuture<CreatingExecutionGraph.ExecutionGraphWithVertexParallelism>
+                    executionGraphWithVertexParallelismFuture = new CompletableFuture<>();
+            final AtomicReference<GlobalFailureHandler> operatorCoordinatorGlobalFailureHandlerRef =
+                    new AtomicReference<>();
+            new CreatingExecutionGraph(
+                    context,
+                    executionGraphWithVertexParallelismFuture,
+                    log,
+                    (executionGraph, errorHandler) -> {
+                        operatorCoordinatorGlobalFailureHandlerRef.set(errorHandler);
+                        return new TestingOperatorCoordinatorHandler();
+                    });
+
+            final StateTrackingMockExecutionGraph executionGraph =
+                    new StateTrackingMockExecutionGraph();
+
+            context.setTryToAssignSlotsFunction(CreatingExecutionGraphTest::successfulAssignment);
+            context.setExpectedExecuting(
+                    ignored -> {
+                        // This just lets us transition to executing.
+                    });
+
+            final AtomicBoolean contextGlobalFailureHandlerCalled = new AtomicBoolean(false);
+            context.setGlobalFailureHandler(
```

Comment:
Couldn't we just check that the errorHandler passed to the factory == context?

Reply:
I'd rather avoid that, because we can't do an assertion in factory as it's executed in the future callback that is wrapped in `FutureUtils.assertNoException`. I don't really want to go into changing any more logic because of that.

I can still make an assert on the reference, but that doesn't seem to make too much of a difference for readability.

WDYT?